thrust/system/detail/generic/shuffle.inl

```diff
@@ -32,6 +32,44 @@ namespace system {
 namespace detail {
 namespace generic {
 
+class round_function {
```

Comment:
I don't like using something like wyhash without a reference. I would prefer using some xor combination of internal thrust random engines from here: https://thrust.github.io/doc/group__predefined__random.html. These engines all have academic references, and we can also reference numerical recipes which recommends combining different types of engines as best practice.

The downside is that this might be a bit slower, but it could be more justified.

Reply:
Fair enough, the difficulty here is actually generating sufficient diffusion. The important feature of this function is that one bit flip causes a large change in the output. This is a feature of hash functions, not random number generators. Perhaps a more well known hash function such as crc would be acceptable?

Reply:
Perhaps https://blake2.net/

Reply:
Additionally, many of the good random number generators are expensive to construct since they are intended to be create once call many

Reply:
I tested crc32c but it performed very poorly, which is to be expected somewhat since it is a checksum not a hash function. Additionally it is only 32 bit

Reply:
Ok I did some brief research
https://doi.org/10.1002/spe.2179 and concludes that MurmurHash2 has good distribution - particularly the section on avalanche affect.
Additionally, https://softwareengineering.stackexchange.com/questions/49550/which-hashing-algorithm-is-best-for-uniqueness-and-speed provides some benchmarks for a number of functions and also suggests (albeit much less scientifically) that MurmurHash2 has strong randomness.
It seems reasonably fast since we dont need to worry about the overhead of being cryptographically secure and we dont particularly care about collision resistance properties, unless they are very common for close keys.

Reply:
I have pushed a commit with the changes to MurmurHash2. I believe this is slightly weaker than WyHashv4 but it is still much stronger than a random number generator would be. Currently both fail the tests due to insufficient randomness. To fix them we may need to increase the number of rounds, as well as use a stronger random number generator as input

Reply:
> The important feature of this function is that one bit flip causes a large change in the output. This is a feature of hash functions, not random number generators.

Not sure if that's true, consecutive seeds for random number generators should absolutely be uncorrelated. I think the key difference for the hash functions is that they are designed for larger inputs. For our purposes 64 -> 64 bits is fine, so I'm not convinced we need anything other than a good random number generator.

Reply:
Definitely increase the number of rounds though, even if we have to go up to 16+.

Reply:
>Not sure if that's true, consecutive seeds for random number generators should absolutely be uncorrelated. I think the key difference for the hash functions is that they are designed for larger inputs. For our purposes 64 -> 64 bits is fine, so I'm not convinced we need anything other than a good random number generator.

This maybe the case for cryptographic DRBGs, but not necessarily for general random number generator. The diffusion must come from somewhere, either through a very expensive initialisation procedure such as for mersenne twister, which may possibly involve running the prng many times. Alternatively, the random number generator may use a hash function to add diffusion, but then you might as well just use the hash function.
The difference here is that a random number generator such as a LFSR will diverge quickly but will start out very similar if the register is initialised similar, but will still output a similar sequence for the first few values, which is especially concerning since we only ever use the first 8 bits generated for an array of length 256.
One thing to note with the previous implementation is that since the key was applied after the PRNG, the output would always be the same for the item with the corresponding index at a given round. This could potentially lead to previous XOR's being reversed causing each value to only be XOR'd with the key, limiting the total randomness